[BUG] PerformanceMonitorInstaller-2.2.0 failed and deleted the existing PerformanceMonitor database

### Component

Installer (GUI)

### Performance Monitor Version

2.2.0

### SQL Server Version

Microsoft SQL Server 2019 (RTM-CU32) (KB5054833) - 15.0.4430.1 (X64)

### Windows Version

Windows server 2016 (on the sql box)

### Describe the Bug

Not sure if it counts as a bug, but here is what happend:
Ran the install via the GUI for version 2.2.0. It reported before install that I had version 1.3.0 installed (this was wrong, as I had version 2.1.0 installed. This was due to the previous bug where the version number would not be updated when using the GUI installer.)

Started the install process, and after running for some time, it failed. After this, the entire PerformanceMonitor database had been deleted. (I did not chose clean install)

### Steps to Reproduce

Open the GUI installer and run it for a server where a previous install had been done using the GUI installer.

### Expected Behavior

Installation success and excisint data compressed.

### Actual Behavior

Installation failed, PerformanceMonitor database gone.

### Error Messages / Log Output

```text
================================================================================
Performance Monitor Installation Report
================================================================================

INSTALLATION SUMMARY
--------------------------------------------------------------------------------
Status:              FAILED
Start Time:          2026-03-12 20:44:03
End Time:            2026-03-12 20:44:30
Duration:            26,5 seconds
Files Executed:      1
Files Failed:        1

SERVER INFORMATION
--------------------------------------------------------------------------------
Server Name:         SERVERNAME-HERE
SQL Server Edition:  Enterprise Edition (64-bit)

SQL Server Version:
  Microsoft SQL Server 2019 (RTM-CU32) (KB5054833) - 15.0.4430.1 (X64)
  Feb 21 2025 17:28:26
  Copyright (C) 2019 Microsoft Corporation
  Enterprise Edition (64-bit) on Windows Server 2016 Standard 10.0 <X64> (Build 14393: ) (Hypervisor)

INSTALLER INFORMATION
--------------------------------------------------------------------------------
Installer Version:   2.2.0+8ed7a95c6484fda201bfe6f0c56931dc28091a50
Working Directory:   D:\PerformanceMonitor\PerformanceMonitorInstaller-2.2.0
Machine Name:        APP-SERVERNAME
User Name:           my_username

ERRORS
--------------------------------------------------------------------------------
File: 01_install_database.sql
Error: Batch 2 failed:
/*
██████╗ ███████╗██████╗ ███████╗ ██████╗ ██████╗ ███╗   ███╗ █████╗ ███╗   ██╗ ██████╗███████╗
██╔══██╗██╔════╝██╔══██╗██╔════╝██╔═══██╗██╔══██╗████╗ ████║██╔══██╗████╗  ██║██╔════╝██╔════╝
██████╔╝█████╗  ██████╔╝█████╗  ██║   ██║██████╔╝██╔████╔██║███████║██╔██╗ ██║██║     █████╗  
██╔═══╝ ██╔══╝  ██╔══██╗██╔══╝  ██║   ██║██╔══██╗██║╚██╔╝██║██╔══██║██║╚██╗██║██║     ██╔══╝  
██║     ███████╗██║  ██║██║     ╚██████╔╝██║  ██║██║ ╚═╝ ██║██║  ██║██║ ╚████║╚██████╗███████╗
...

DETAILED INSTALLATION LOG
--------------------------------------------------------------------------------
Performance Monitor Installer v2.2.0+8ed7a95c6484fda201bfe6f0c56931dc28091a50
Copyright (c) 2026 Darling Data, LLC
Found 3 upgrade(s) to apply
Applying upgrade: 1.3.0-to-2.0.0
[OK]   01_memory_grant_stats_schema.sql - Success
[OK]   02_session_wait_stats_cleanup.sql - Success
[OK]   03_default_trace_schema.sql - Success
[OK] Upgrade 1.3.0-to-2.0.0: 3 succeeded, 0 failed
Applying upgrade: 2.0.0-to-2.1.0
[OK]   01_collection_schedule_optional_text.sql - Success
[OK]   02_default_trace_events_new_columns.sql - Success
[OK] Upgrade 2.0.0-to-2.1.0: 2 succeeded, 0 failed
Applying upgrade: 2.1.0-to-2.2.0
[ERROR]   01_compress_query_stats.sql - FAILED: Batch 6 failed:
BEGIN TRY

    /*
    Step 1: Create the _new table with compressed column types
    */
    IF OBJECT_ID(N'collect.query_stats_new', N'U') IS NOT NULL
    BEGIN
        DROP TABLE collect.query_stats_new;
        PRINT 'Dropped existing collect.query_stats_new';
    END;

    CREATE TABLE
        collect.query_stats_new
    (
        collection_id bigint IDENTITY NOT NULL,
        collection_time datetime2(7) NOT NULL
            DEFAULT SYSDATETIME(),
        server_start_time ... [truncated, total length: 11086]

Original error: Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
Operation cancelled by user.
Created collect.query_stats_new
Checking identity information: current identity value 'NULL'.
DBCC execution completed. If DBCC printed error messages, contact your system administrator.
Reseeded IDENTITY to 2656731
 
Migrating data in batches of 10000...
  Migrated 10000 rows so far...
  Migrated 20000 rows so far...
  Migrated 30000 rows so far...
  Migrated 40000 rows so far...
  Migrated 50000 rows so far...
  Migrated 60000 rows so far...
  Migrated 70000 rows so far...
  Migrated 80000 rows so far...
  Migrated 90000 rows so far...
  Migrated 100000 rows so far...
  Migrated 110000 rows so far...
  Migrated 120000 rows so far...
The statement has been terminated.
[OK]   02_compress_query_store_data.sql - Success
[OK]   03_compress_procedure_stats.sql - Success
[OK]   04_create_tracking_tables.sql - Success
[OK]   05_add_finops_collectors.sql - Success
[WARN] Upgrade 2.1.0-to-2.2.0: 4 succeeded, 1 failed
Starting installation...
Executing 00_uninstall.sql...
[OK] 00_uninstall.sql - Success
Executing 01_install_database.sql...
[ERROR] 01_install_database.sql - FAILED: Batch 2 failed:
/*
██████╗ ███████╗██████╗ ███████╗ ██████╗ ██████╗ ███╗   ███╗ █████╗ ███╗   ██╗ ██████╗███████╗
██╔══██╗██╔════╝██╔══██╗██╔════╝██╔═══██╗██╔══██╗████╗ ████║██╔══██╗████╗  ██║██╔════╝██╔════╝
██████╔╝█████╗  ██████╔╝█████╗  ██║   ██║██████╔╝██╔████╔██║███████║██╔██╗ ██║██║     █████╗  
██╔═══╝ ██╔══╝  ██╔══██╗██╔══╝  ██║   ██║██╔══██╗██║╚██╔╝██║██╔══██║██║╚██╗██║██║     ██╔══╝  
██║     ███████╗██║  ██║██║     ╚██████╔╝██║  ██║██║ ╚═╝ ██║██║  ██║██║ ╚████║╚██████╗███████╗
╚═╝     ╚══════╝... [truncated, total length: 3612]

Original error: Could not obtain exclusive lock on database 'model'. Retry the operation later.
CREATE DATABASE failed. Some file names listed could not be created. Check related errors.
User does not have permission to alter database 'PerformanceMonitor', the database does not exist, or the database is not in a state that allows access checks.
ALTER DATABASE statement failed.
Created PerformanceMonitor database at d:\data\PerformanceMonitor.mdf
[ERROR] Critical installation file failed. Aborting installation.

================================================================================
Generated by Performance Monitor Installer GUI
Copyright (c) 2026 Darling Data, LLC
================================================================================
```

### Screenshots

<img width="1218" height="989" alt="Image" src="https://github.com/user-attachments/assets/72e8b1d9-ee0f-4ba1-8f2c-47ba7ef7c9c1" />

### Additional Context

The existing PerformanceMonitor database was pretty big (about 240 GB).
I tried this install on a testserver with a very small PerformanceMonitor database before this, and that installed normally.

--- a/Installer/Program.cs
+++ b/Installer/Program.cs
@@ -103,6 +103,7 @@ END;";
         private const int ShortTimeoutSeconds = 60;       // Quick operations (cleanup, queries)
         private const int MediumTimeoutSeconds = 120;     // Dependency installation
         private const int LongTimeoutSeconds = 300;       // SQL file execution (5 minutes)
+        private const int UpgradeTimeoutSeconds = 3600;   // Upgrade data migrations (1 hour, large tables)
 
         /*
         Exit codes for granular error reporting
@@ -117,6 +118,7 @@ END;";
             public const int VersionCheckFailed = 5;
             public const int SqlFilesNotFound = 6;
             public const int UninstallFailed = 7;
+            public const int UpgradesFailed = 8;
         }
 
         static async Task<int> Main(string[] args)
@@ -523,8 +525,9 @@ END;";
                     string fileName = Path.GetFileName(f);
                     if (!SqlFileNamePattern.IsMatch(fileName))
                         return false;
-                    /*Exclude test and troubleshooting scripts from main install*/
-                    if (fileName.StartsWith("97_", StringComparison.Ordinal) ||
+                    /*Exclude uninstall, test, and troubleshooting scripts from main install*/
+                    if (fileName.StartsWith("00_", StringComparison.Ordinal) ||
+                        fileName.StartsWith("97_", StringComparison.Ordinal) ||
                         fileName.StartsWith("99_", StringComparison.Ordinal))
                         return false;
                     return true;
@@ -699,6 +702,21 @@ END;";
 
                         Console.WriteLine();
                         Console.WriteLine($"Upgrades complete: {upgradeSuccessCount} succeeded, {upgradeFailureCount} failed");
+
+                        /*Abort if any upgrade scripts failed — proceeding would reinstall over a partially-upgraded database*/
+                        if (upgradeFailureCount > 0)
+                        {
+                            Console.WriteLine();
+                            Console.WriteLine("================================================================================");
+                            Console.WriteLine("Installation aborted: upgrade scripts must succeed before installation can proceed.");
+                            Console.WriteLine("Fix the errors above and re-run the installer.");
+                            Console.WriteLine("================================================================================");
+                            if (!automatedMode)
+                            {
+                                WaitForExit();
+                            }
+                            return ExitCodes.UpgradesFailed;
+                        }
                     }
                     else
                     {
@@ -1332,7 +1350,15 @@ END;";
                         return version.ToString();
                     }
 
-                    return null;
+                    /*
+                    Fallback: database and history table exist but no SUCCESS rows.
+                    This can happen if a prior GUI install didn't write history (#538/#539).
+                    Return "1.0.0" so all idempotent upgrade scripts are attempted
+                    rather than treating this as a fresh install (which would drop the database).
+                    */
+                    Console.WriteLine("Warning: PerformanceMonitor database exists but installation_history has no records.");
+                    Console.WriteLine("Treating as v1.0.0 to apply all available upgrades.");
+                    return "1.0.0";
                 }
             }
             catch (SqlException ex)
@@ -1480,7 +1506,7 @@ END;";
 
                         using (var cmd = new SqlCommand(trimmedBatch, connection))
                         {
-                            cmd.CommandTimeout = LongTimeoutSeconds;
+                            cmd.CommandTimeout = UpgradeTimeoutSeconds;
                             try
                             {
                                 await cmd.ExecuteNonQueryAsync().ConfigureAwait(false);

--- a/InstallerGui/MainWindow.xaml.cs
+++ b/InstallerGui/MainWindow.xaml.cs
@@ -408,6 +408,16 @@ namespace PerformanceMonitorInstallerGui
                             upgradeFailure == 0 ? "Success" : "Warning");
                         LogMessage("", "Info");
                     }
+
+                    /*Abort if any upgrade scripts failed — proceeding would reinstall over a partially-upgraded database*/
+                    if (upgradeFailure > 0)
+                    {
+                        LogMessage("", "Info");
+                        LogMessage("Installation aborted: upgrade scripts must succeed before installation can proceed.", "Error");
+                        LogMessage("Fix the errors above and re-run the installer.", "Error");
+                        SetUIState(installing: false);
+                        return;
+                    }
                 }
 
                 /*

--- a/InstallerGui/Services/InstallationService.cs
+++ b/InstallerGui/Services/InstallationService.cs
@@ -238,8 +238,9 @@ namespace PerformanceMonitorInstallerGui.Services
                         /*Match numbered SQL files but exclude 97 (tests) and 99 (troubleshooting)*/
                         if (!SqlFilePattern.IsMatch(fileName))
                             return false;
-                        /*Exclude test and troubleshooting scripts from main install*/
-                        if (fileName.StartsWith("97_", StringComparison.Ordinal) ||
+                        /*Exclude uninstall, test, and troubleshooting scripts from main install*/
+                        if (fileName.StartsWith("00_", StringComparison.Ordinal) ||
+                            fileName.StartsWith("97_", StringComparison.Ordinal) ||
                             fileName.StartsWith("99_", StringComparison.Ordinal))
                             return false;
                         return true;
@@ -1113,7 +1114,13 @@ END;";
                     return version.ToString();
                 }
 
-                return null;
+                /*
+                Fallback: database and history table exist but no SUCCESS rows.
+                This can happen if a prior GUI install didn't write history (#538/#539).
+                Return "1.0.0" so all idempotent upgrade scripts are attempted
+                rather than treating this as a fresh install (which would drop the database).
+                */
+                return "1.0.0";
             }
             catch (SqlException)
             {
@@ -1272,7 +1279,7 @@ END;";
                             continue;
 
                         using var cmd = new SqlCommand(trimmedBatch, connection);
-                        cmd.CommandTimeout = 300;
+                        cmd.CommandTimeout = 3600; /*1 hour — upgrade migrations on large tables need extended time*/
 
                         try
                         {

